projects/igniteui-angular/src/lib/grids/pivot-grid/pivot-util.ts

```diff
@@ -0,0 +1,476 @@
+                }
+            }
+        }
+        return hierarchy;
+    }
+
+    public static sort(data: any[], expressions: ISortingExpression[], sorting: IGridSortingStrategy = new IgxSorting(), pivotKeys): any[] {
+        data.forEach(rec => {
+            const keys = Object.keys(rec);
+            rec.sorted = true;
+            keys.forEach(k => {
+                if (k.indexOf(pivotKeys.records) !== -1 && k !== pivotKeys.records) {
+                    const unsorted = rec[k].filter(x => !x.sorted);
+                    // sort all child record collections based on expression recursively.
+                    rec[k] = this.sort(unsorted, expressions, sorting, pivotKeys);
+                }
+            });
+        });
+        return DataUtil.sort(data, expressions, sorting);
+    }
+
+    public static extractValueFromDimension(dim: IPivotDimension, recData: any) {
+        return dim.memberFunction ? dim.memberFunction.call(null, recData) : recData[dim.memberName];
+    }
+
+    public static getDimensionDepth(dim: IPivotDimension): number {
+        let lvl = 0;
+        while (dim.childLevel) {
+            lvl++;
+            dim = dim.childLevel;
+        }
+        return lvl;
+    }
+
+    public static getDimensionLevel(dim: IPivotDimension, rec: any, pivotKeys: IPivotKeys) {
+        let level = rec[dim.memberName + pivotKeys.rowDimensionSeparator + pivotKeys.level];
+        while (dim.childLevel && level === undefined) {
+            dim = dim.childLevel;
+            level = rec[dim.memberName + pivotKeys.rowDimensionSeparator + pivotKeys.level];
+        }
+        return { level, dimension: dim };
+    }
+
+    public static flattenHierarchy(records: any[],
+        config: IPivotConfiguration,
+        dim: IPivotDimension,
+        expansionStates: any,
+        defaultExpandState: boolean,
+        pivotKeys: IPivotKeys,
+        lvl: number,
+        prevDims: IPivotDimension[],
+        currDimLvl: number,
+        maxDimLvl: number) {
+        const data = records;
+        for (let i = 0; i < data.length; i++) {
+            const rec = data[i];
+            const field = dim.memberName;
+            if (!field) {
+                continue;
+            }
+            rec[field + pivotKeys.rowDimensionSeparator + pivotKeys.level] = currDimLvl;
+            const expansionRowKey = PivotUtil.getRecordKey(rec, dim, prevDims, pivotKeys);
+            const isExpanded = expansionStates.get(expansionRowKey) === undefined ?
+                defaultExpandState :
+                expansionStates.get(expansionRowKey);
+            if (rec[field + pivotKeys.rowDimensionSeparator + pivotKeys.records] &&
+                rec[field + pivotKeys.rowDimensionSeparator + pivotKeys.records].length > 0 &&
+                ((isExpanded && lvl > 0) || (maxDimLvl == currDimLvl))) {
+                let dimData = rec[field + pivotKeys.rowDimensionSeparator + pivotKeys.records];
```

Comment:
Yeah I think it could be beneficial. But I'll let @rkaraivanov chime in on that as well.

Reply:
Main concern if we want to change it would be that the pipes that would generate this PivotRecord are the row/column pipes since they build the hierarchy and aggregations. They can be skipped or can be custom (if user uses remote data that is pre-processed for example), while the expand/collapse pipe is non-skippable and is always resolved on the client. So if we make all pipes operated based on this PivotRecord object and an user wants to pass a pre-processed json data, he would have to transform it to a PivotRecords before it can be consumed by the grid or we would have to make another intermediate pipe or method that can do it for him.